submissions/crealime/oop-exercise/index.js

```diff
@@ -0,0 +1,143 @@
+    return `${super.getFullName()} ${this.surname}`
+  }
+}
+
+class CatWoman extends Inhabitant {
+  constructor(name, surname) {
+    super('cat-woman', name, 'female', Cat.say())
+    this.surname = surname
+    this.legs = 2
+    this.hands = 2
+  }
+
+  getFullName() {
+    return `${this.name} ${this.surname}`
+  }
+}
+
+const dog = new Dog('Marley', 'male', 4)
+const cat = new Cat('Sophie', 'female', 4)
+const man = new Human('Harry', 'Taylor', 'male', 2, 2, 'Hello everyone!')
+const woman = new Human('Emily', 'Brown', 'female', 2, 2, 'Hi all!')
+const pirate = new Human('John', 'Silver', 'male', 1, 2, 'Give no quarter!')
+const catWoman = new CatWoman('Selina', 'Kyle')
+
+dog.friends.push(cat, man)
+cat.friends.push(dog, woman, catWoman)
+man.friends.push(dog, woman)
+woman.friends.push(cat, man)
+catWoman.friends.push(cat)
+
+const secretSociety = []
+const allInhabitant = [dog, cat, man, woman, catWoman, pirate]
+
+function addToSociety(society, member) {
+  society.push(member.id)
+}
+
+addToSociety(secretSociety, catWoman)
+addToSociety(secretSociety, pirate)
+
+function isInSociety(society, members) {
+  return members.filter(member => society.includes(member.id)).map(member => member.name)
+}
+
+// ======== OUTPUT ========
+/* Use print(message) for output.
+   Default tag for message is <pre>. Use print(message,'div') to change containing element tag.
+
+   Message can contain HTML markup. You may also tweak index.html and/or styles.css.
+   However, please, REFRAIN from improving visuals at least until your code is reviewed
+   so code reviewers might focus on a single file that is index.js.
+   */
+
+/* Print examples:
+   print('ABC');
+   print('<strong>ABC</strong>');
+   print('<strong>ABC</strong>', 'div');
+
+   print('human; John; male; 2; 2; Hello world!; Rex, Tom, Jenny');
+   print('human; <strong>John</strong>; male; 2; 2; <em>Hello world!</em>; Rex, Tom, Jenny');
+   print('human; <strong>John</strong>; male; 2; 2; <em>Hello world!</em>; Rex, Tom, Jenny', 'div');
+   */
+
+function getFullInfo(inhabitant) {
+  const getAcquaintance = `This is ${inhabitant.getFullName()} and ${inhabitant.gender === 'male' ? 'he' : 'she'} is a ${inhabitant.species}.`
+  const getGender = `${inhabitant.gender === 'male' ? 'His' : 'Her'} gender is ${inhabitant.gender}.`
+  const getLimbs = `${inhabitant.gender === 'male' ? 'He' : 'She'} have ${inhabitant.hands || 0} ${inhabitant.hands === 1 ? 'hand' : 'hands'} and ${inhabitant.legs || 0} ${inhabitant.legs === 1 ? 'leg' : 'legs'}.`
+  const getFriends = `${inhabitant.name} have ${inhabitant.friends.length} ${inhabitant.friends.length === 1 ? 'friend' : 'friends'}${inhabitant.friends.length > 0 ? ': ' + inhabitant.friends.map(friend => friend.name).join(', ') : ''}.`
+  const getSaying = `${inhabitant.gender === 'male' ? 'He' : 'She'} usually says hello like this: ${inhabitant.saying}`
+
+  return `${getAcquaintance} ${getGender} ${getLimbs} ${getFriends} ${getSaying}`
+}
```

Comment:
Are there any good reasons not have all this handled by classes? (not a rhetoric question)
Imagine, we add fish, birds, octopi, and spiders. How big this function would become? (a rhetoric question)
Separation of concerns and encapsulation are good principles to follow.

There must be good reasons to handle full name in one way and other props in a completely different way. What these reasons are?

Reply:
Yes, at first I thought that not all residents can tell about themselves. Then I decidet that there can be many such functions in the world, some functions can be created dynamically. Therefore, I did not put it in classes. This feature is quite flexible, but it can be added functionality for those inhabitants, who do not make sounds at all, such as fish. Мaybe gender for hermaphrodites.. However, to practice, I added this mixin to Human and CatWoman classes. 

I made getFullName method because some inhabitants  have a first and last name, while others have only first name. I thought it was more useful.

Reply:
Composition would work like this. Mixing OOP and composition is not the best practice.
`getFullName` is valid case though. And this is the key -- when behaviour in a subclass changes, we overload the method. Re-use this approach.
Composition is more flexible approach. OOP, from the other hand, helps to encapsulate the logic in a way that every class handles its own properties. So, for e.g. Humans, they add only info on hands on top of presentation prepared by the base class.
Please refactor in OOP-compliant manner

Reply:
> This feature is quite flexible, but it can be added functionality for those inhabitants, who do not make sounds at all, such as fish. Мaybe gender for hermaphrodites.

We design classes in a manner that the code is easily extensible, yet we do not make pre-mature optimizations. SImply, because we cannot really predict what would be required.